test/unit/ipam_test.rb

```diff
@@ -0,0 +1,90 @@
+require 'test_helper'
+
+class IPAMTest < ActiveSupport::TestCase
+  test 'should raise an exception if given mac is invalid' do
+    s = FactoryGirl.build(:subnet_ipv6, :network => '2001:db8::')
+    exception = assert_raise Foreman::Exception do
+      IPAM::Eui64.new(:subnet => s, :mac => 'invalid')
+    end
+    assert_includes exception.message, 'not a valid MAC address'
+  end
+
+  context 'dhcp' do
+    test "should find unused IP on proxy if proxy is set" do
+      subnet = FactoryGirl.create(:subnet_ipv4, :ipam_dhcp, :name => 'my_subnet', :network => '192.168.1.0')
+      subnet.stubs(:dhcp? => true)
+      subnet.stubs(:dhcp => mock('attribute', :url => 'proxy.example.com'))
+      fake_proxy = mock("dhcp_proxy")
+      fake_proxy.stubs(:unused_ip => {'ip' => '192.168.1.25'})
+      subnet.stubs(:dhcp_proxy => fake_proxy)
+      ipam = IPAM::Dhcp.new(:subnet => subnet, :mac => '00:11:22:33:44:55')
+      assert_equal '192.168.1.25', ipam.suggest_ip
+    end
+  end
+
+  context 'internal db' do
+    test "should find unused IP in internal DB if proxy is not set" do
+      host = FactoryGirl.create(:host)
```

Comment:
Actually I think this is testing too much. By creating hosts, we're testing `.known_ips` and stuff we shouldn't be testing here. [That's tested somewhere else already](https://github.com/theforeman/foreman/blob/develop/test/unit/subnet/ipv4_test.rb#L287). We just want to get a free IP from the db. So two options to offload this test are: 

1 - passing `:excluded_ips` to the IPAM service.
2 - stubbing `subnet.known_ips` to return whatever you want.

It'd look like this, you'd save creating an expensive object in db, and also it's much clearer. To understand the current test now one has to check the implementation.

```ruby
test do
  subnet = FactoryGirl.create(
          :subnet_ipv4, :name => 'my_subnet',
          :network => '192.168.2.0',
          :ipam => IPAM::MODES[:db])
  ipam = IPAM::Db.new(:subnet => subnet, :excluded_ips => ['192.168.2.1', '192.168.2.2'])
  assert_equal '192.168.2.3', ipam.suggest_ip
end 
```